mmn/collector.go

```diff
@@ -110,41 +113,64 @@ func hashMACAddress(mac string) string {
 func getOS() string {
 	switch runtime.GOOS {
 	case "darwin":
-		output, err := shell.NewCmd("sw_vers -productVersion").Output()
+		output, err := exec.Command("sw_vers", "-productVersion").Output()
 		if err != nil {
-			log.Error().Err(err).Msg("Failed to get OS information")
-			return ""
+			log.Err(err).Msg("Failed to get OS information")
+			return "macOS (unknown)"
 		}
-		return "MAC OS X - " + strings.TrimSpace(string(output))
-
+		return "macOS " + strings.TrimSpace(string(output))
 	case "linux":
-		output, err := shell.NewCmd("lsb_release -d").Output()
+		distro, err := parseLinuxOS()
 		if err != nil {
-			output, err = shell.NewCmd("source /etc/os-release && echo $PRETTY_NAME").Output()
-			if err != nil {
-				log.Error().Err(err).Msg("Failed to get OS information")
-				return ""
-			}
+			log.Err(err).Msg("Failed to get OS information")
+			return "linux (unknown)"
 		}
-		return strings.TrimSpace(strings.Replace(string(output), "Description:", "", 1))
-
+		return distro
 	case "windows":
-		output, err := shell.NewCmd("wmic os get Caption /value").Output()
+		output, err := exec.Command("wmic", "os", "get", "Caption", "/value").Output()
 		if err != nil {
-			log.Error().Err(err).Msg("Failed to get OS information")
-			return ""
+			log.Err(err).Msg("Failed to get OS information")
+			return "windows (unknown)"
 		}
-		return extractWindowsVersion(string(output))
+		return strings.TrimSpace(strings.TrimPrefix(string(output), "Caption="))
 	}
-
-	return ""
+	return runtime.GOOS
 }
 
-func extractWindowsVersion(output string) string {
-	var version string
+func parseLinuxOS() (string, error) {
+	output, err := exec.Command("lsb_release", "-d").Output()
+	if err == nil {
+		return strings.TrimSpace(strings.TrimPrefix(string(output), "Description:")), nil
+	}
 
-	version = strings.TrimSpace(strings.Replace(output, "Caption=", "", 1))
-	version = strings.TrimSpace(strings.Replace(version, "Microsoft", "", 1))
+	const etcOsRelease = "/etc/os-release"
+	const altOsRelease = "/usr/lib/os-release"
+	osReleaseFile, err := os.Open(etcOsRelease)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return "", fmt.Errorf("error opening %s: %w", etcOsRelease, err)
+		}
+		osReleaseFile, err = os.Open(altOsRelease)
+		if err != nil {
+			return "", fmt.Errorf("error opening %s: %w", altOsRelease, err)
+		}
+	}
+	defer osReleaseFile.Close()
+
+	var prettyName string
+	scanner := bufio.NewScanner(osReleaseFile)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, "PRETTY_NAME") {
+			tokens := strings.SplitN(line, "=", 2)
+			if len(tokens) == 2 {
+				prettyName = strings.Trim(tokens[1], "\"")
```

Comment:
At this point you can stop loop.

```go
prettyName = strings.Trim(tokens[1], "\"")
break
```

Reply:
Yes, but it doesn’t matter as the file is tiny. So I’m intentionally reducing complexity here

Reply:
Why should someone care if file is tiny. Breaking the loop when you done is also much more clear and indicates that after name is one no more cycles are needed.

Reply:
Because every control instruction, code branch, etc. increases complexity for the reader and since there are only a few cycles to be wasted, that is the rationale for not accounting for them (as already explained). Here, for instance, is the same logic implemented in docker core: https://github.com/moby/moby/blob/master/pkg/parsers/operatingsystem/operatingsystem_linux.go#L64

Reply:
I guess it's because value can be overwritten so in this case the last one will be taken.